File: system-test/java/src/main/java/com/mycompany/myshop/systemtest/configuration/PropertyLoader.java
package com.mycompany.myshop.systemtest.configuration;

import com.mycompany.myshop.testkit.dsl.port.ChannelMode;

public class PropertyLoader {
    private PropertyLoader() {
    }

    public static Environment getEnvironment(Environment fixedEnvironment) {
        if (fixedEnvironment != null) {
            return fixedEnvironment;
        }

        // Resolve in order: -Denvironment system property (used by JSON commands like
        // gradle test -Denvironment=local), then ENVIRONMENT env var (set at workflow
        // shell level for acceptance/qa stages), then default LOCAL. This lets one
        // tests-*.json drive every environment without per-suite -D wiring.
        var fromProperty = System.getProperty("environment");
        if (fromProperty != null && !fromProperty.isBlank()) {
            return Environment.valueOf(fromProperty.toUpperCase());
        }
        var fromEnv = System.getenv("ENVIRONMENT");
        if (fromEnv != null && !fromEnv.isBlank()) {
            return Environment.valueOf(fromEnv.toUpperCase());
        }
        return Environment.LOCAL;
    }

    public static ExternalSystemMode getExternalSystemMode(ExternalSystemMode fixedExternalSystemMode) {
        if (fixedExternalSystemMode != null) {
            return fixedExternalSystemMode;
        }

        var externalSystemMode = getRequiredSystemProperty("externalSystemMode", "stub|real");
        return ExternalSystemMode.valueOf(externalSystemMode.toUpperCase());
    }

    public static ChannelMode getChannelMode(ChannelMode fixedChannelMode) {
        if (fixedChannelMode != null) {
            return fixedChannelMode;
        }

        var value = System.getProperty("channelMode");
        if (value == null || value.isBlank()) {
            return ChannelMode.DYNAMIC;
        }
        return ChannelMode.valueOf(value.toUpperCase());
    }

    private static String getRequiredSystemProperty(String propertyName, String allowedValues) {
        var value = System.getProperty(propertyName);
        if (value == null || value.isBlank()) {
            throw new IllegalStateException(
                    String.format("System property '%s' is not defined. Please specify -D%s=<%s>",
                            propertyName, propertyName, allowedValues)
            );
        }
        return value;
    }
}

File: system-test/dotnet/SystemTests/TestInfrastructure/Configuration/PropertyLoader.cs
using Dsl.Port;
using Dsl.Core.Shared;

namespace SystemTests.TestInfrastructure.Configuration;

public static class PropertyLoader
{
    public static Environment GetEnvironment(Environment? fixedEnvironment)
    {
        if (fixedEnvironment != null)
        {
            return fixedEnvironment.Value;
        }

        // Resolve from ENVIRONMENT env var, defaulting to Local. The workflow drives
        // the value via shell env (acceptance/qa stages) and the JSON tests-*.json
        // commands no longer hardcode -e ENVIRONMENT=local — see local stage and
        // cross-lang, which both rely on this default.
        var environmentMode = System.Environment.GetEnvironmentVariable("ENVIRONMENT");
        return string.IsNullOrEmpty(environmentMode)
            ? Environment.Local
            : Enum.Parse<Environment>(environmentMode, ignoreCase: true);
    }

    public static ExternalSystemMode GetExternalSystemMode(ExternalSystemMode? fixedExternalSystemMode)
    {
        if (fixedExternalSystemMode != null)
        {
            return fixedExternalSystemMode.Value;
        }

        var externalSystemMode = GetRequiredEnvironmentVariable("EXTERNAL_SYSTEM_MODE", "stub|real");
        return Enum.Parse<ExternalSystemMode>(externalSystemMode, ignoreCase: true);
    }

    public static ChannelMode GetChannelMode(ChannelMode? fixedChannelMode)
    {
        if (fixedChannelMode != null)
        {
            return fixedChannelMode.Value;
        }

        var value = System.Environment.GetEnvironmentVariable("CHANNEL_MODE");
        if (string.IsNullOrEmpty(value))
        {
            return ChannelMode.Dynamic;
        }
        return Enum.Parse<ChannelMode>(value, ignoreCase: true);
    }

    private static string GetRequiredEnvironmentVariable(string variableName, string allowedValues)
    {
        var value = System.Environment.GetEnvironmentVariable(variableName);

        if (string.IsNullOrEmpty(value))
        {
            throw new InvalidOperationException(
                $"Environment variable '{variableName}' is not defined. Please set {variableName}=<{allowedValues}>");
        }

        return value;
    }
}










